### DISCUSSION TITLE
[General] Getting nonce error when interacting with contract

### DISCUSSION THREAD

**Discussion by abhishek-Kumar009** [zkSync-Community-Hub/zksync-developers#419] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

I deployed Uniswap V2 factory and router on Zksync era mainnet by forking it. On using addLiquidity of Uniswap V2 router I'm getting nonce error

### What is the specific issue or error you're encountering?

Nonce error

### Can you share the error messages or logs you're receiving, if any?

`14:04:35  INFO Validating 0xffe5f8b36e3c326fd2f92f0aeb2d75dd9350d62443ae9a4209835e1531949264
14:04:35  INFO Executing 0xffe5f8b36e3c326fd2f92f0aeb2d75dd9350d62443ae9a4209835e1531949264
14:04:36  INFO ┌─────────────────────────┐
14:04:36  INFO │   TRANSACTION SUMMARY   │
14:04:36  INFO └─────────────────────────┘
14:04:36  INFO Transaction: HALTED
14:04:36  INFO Initiator: 0xd986b0cb0d1ad4cccf0c4947554003fc0be548e9
14:04:36  INFO Payer: 0xd986b0cb0d1ad4cccf0c4947554003fc0be548e9
14:04:36  INFO Gas - Limit: 1_000_000 | Used: 1_000_000 | Refunded: 0
14:04:36  INFO Use --show-gas-details flag or call config_setShowGasDetails to display more info
14:04:36  INFO 
14:04:36  INFO ==== Console logs: 
14:04:36  INFO 
14:04:36  INFO ==== 10 call traces.  Use --show-calls flag or call config_setShowCalls to display more info.
14:04:36  INFO Call(Normal) Account Code Storage                                 getRawCodeHash(address) Revert: Error function_selector = 0x0108c379, data = 0x0108c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001c52657573696e67207468652073616d65206e6f6e636520747769636500000000  4227857550
14:04:38  INFO   Call(Normal) System context                                             0x02fa5779   4227855030
14:04:38  INFO     Call(Normal) Keccak                                               00000000   4161793734
14:04:38  INFO   Call(Normal) Account Code Storage                                 getRawCodeHash(address)   4227687639
14:04:39  INFO   Call(Normal) System context                                             0x06bed036   4227685434
14:04:39  INFO     Call(Normal) Keccak                                               00000000   4161626280
14:04:39  INFO   Call(Normal) System context                                       baseFee()   4227667164
14:04:39  INFO   Call(Normal) Bootloader utilities                                 getTransactionHashes(tuple)   4227657903
14:04:39  INFO     Call(Normal) System context                                       chainId()   4161597867
14:04:39  INFO     Call(Normal) Keccak                                               f9012c15   4161594402
14:04:39  INFO     Call(Normal) System context                                       chainId()   4161589362
14:04:39  INFO     Call(Normal) Keccak                                               f9016c15   4161585204
14:04:40  INFO   Call(Normal) System context                                             0x06e7517b   4227641271
14:04:40  INFO     Call(Normal) Keccak                                               00000000   4161583566
14:04:40  INFO   Call(Normal) System context                                       setTxOrigin(address)   960372
14:04:40  INFO   Call(Normal) System context                                       setGasPrice(uint256)   956088
14:04:40  INFO   Call(Normal) L2 deployer                                          extendedAccountVersion(address)   946449
14:04:40  INFO     Call(Normal) Keccak                                               00000000   931140
14:04:40  INFO     Call(Normal) Account Code Storage                                 getRawCodeHash(address)   930321
14:04:40  INFO   Call(Normal) Nonce Holder                                         validateNonceUsage(address, uint256, bool) Revert: Reusing the same nonce twice  942984
14:04:40  INFO     Call(Normal) Keccak  `

### Have you made any recent changes to the contract before encountering this issue?

NO

### Are there any external libraries or contracts that your contract interacts with?

`    "@matterlabs/hardhat-zksync-deploy": "^1.1.2",
    "@matterlabs/hardhat-zksync-node": "^1.0.1",
    "@matterlabs/hardhat-zksync-solc": "^1.0.6",
    "@matterlabs/hardhat-zksync-verify": "^1.2.2",
    "@matterlabs/zksync-contracts": "^0.6.1",
    "@nomiclabs/hardhat-etherscan": "^3.1.7",
    "@openzeppelin/contracts": "^4.6.0",
    "@types/chai": "^4.3.4",
    "@types/mocha": "^10.0.1",
    "chai": "^4.3.7",
    "dotenv": "^16.0.3",
    "ethers": "^6.9.2",
    "hardhat": "^2.12.4",
    "mocha": "^10.2.0",
    "ts-node": "^10.9.1",
    "typescript": "^4.9.5",
    "zksync-ethers": "^6.0.0"`

### Can you provide the relevant portions of your contract code where the issue is occurring?

`    function addLiquidity(
        address tokenA,
        address tokenB,
        uint amountADesired,
        uint amountBDesired,
        uint amountAMin,
        uint amountBMin,
        address to,
        uint deadline
    ) external virtual override ensure(deadline) returns (uint amountA, uint amountB, uint liquidity) {
        (amountA, amountB) = _addLiquidity(tokenA, tokenB, amountADesired, amountBDesired, amountAMin, amountBMin);
        address pair = UniswapV2Library.pairFor(factory, tokenA, tokenB);
        TransferHelper.safeTransferFrom(tokenA, msg.sender, pair, amountA);
        TransferHelper.safeTransferFrom(tokenB, msg.sender, pair, amountB);
        liquidity = IUniswapV2Pair(pair).mint(to);
    }`

### Have you tried to isolate the problem, and if so, what were the results?

The code flow doesn't reach the function because of this nonce issue as far as i can understand

### What steps have you already taken to try to resolve the issue?

Tried running mainnet node locally and identified the exact issue i.e Nonce

### Repo Link (Optional)

https://github.com/abhishek-Kumar009/Zksync-uniswap

**Comment by abhishek-Kumar009** [comment]
<img width="1344" alt="Screenshot 2024-03-20 at 7 35 23 PM" src="https://github.com/zkSync-Community-Hub/zksync-developers/assets/39706657/94bc72d7-9ea1-44c6-95b9-6c1bb8daef34">

**Comment by uF4No** [comment, marked as answer]
Hi @abhishek-Kumar009 
ethers v6 uses a local cache and in some cases the request to retrieve the next nonce is not sent and it uses the last cached data, which causes the `same nonce error` you're getting. To avoid this, initialise the provider in your tests as follows to disable ethers' cache:

```js
const provider = new Provider("http://127.0.0.1:8011/", undefined, {cacheTimeout: -5});
```

**Reply by abhishek-Kumar009** [reply]
Did this still getting same error
Infact I also tried switching to ethers v5
Same error pops up

**Reply by shonentropy3** [reply]
Maybe you can refer to here https://docs.zksync.io/build/quick-start/best-practices.html#use-call-over-send-or-transfer

**Reply by bxpana** [reply]
@abhishek-Kumar009 we had someone with a similar issue and this is what they did to resolve it:

> Generally initial code hash is used in `pairFor` function in uniswap v2. Hardcoding the initial code hash won't cause an issue in EVM-based chains, but since the compiler of zkSync is different, it is different for zkSync side. So, modifying the init code hash **or** getting pair address by calling `getPair` function of factory contract are needed. Second way consumes a bit gas of course due to external call. https://github.com/Uniswap/v2-periphery/blob/master/contracts/libraries/UniswapV2Library.sol#L24

Can you let us know if that works for you?
